tests/validation/tests/v3_api/test_websocket.py

```diff
@@ -0,0 +1,154 @@
+from threading import Thread
+from .common import CATTLE_TEST_URL
+from .common import create_kubeconfig
+from .common import get_user_client_and_cluster
+from .common import get_project_client_for_token
+from .common import USER_TOKEN
+import websocket
+import ssl
+import base64
+import time
+import pytest
+
+last_message = ""
+namespace = {"cluster": None, "shell_url": None, "pod": None, "ns": ""}
+
+
+def send_a_command(ws_connection, command):
+    cmd_enc = base64.b64encode(command.encode('utf-8')).decode('utf-8')
+    ws_connection.send('0' + cmd_enc)
+    # sends the command to the webSocket
+    ws_connection.send('0DQ==')
+    time.sleep(3)
+
+
+def test_websocket_launch_kubectl():
+    global last_message
+    ws = websocket.create_connection(
+        url=namespace["shell_url"],
+        sslopt={"cert_reqs": ssl.CERT_NONE},
+        subprotocols=["base64.channel.k8s.io"],
+        timeout=10,
+        cookie="R_SESS=" + USER_TOKEN
+    )
+    assert ws.connected, "failed to build the websocket"
+    thread = Thread(target=receiver, args=(ws, True))
+    thread.daemon = True
+    thread.start()
+    time.sleep(1)
+
+    cmd = "kubectl version"
+    send_a_command(ws, cmd)
+    print('kubectl command and output:\n' + last_message + '\n')
+    assert "Client Version" in last_message or \
+           "Server Version" in last_message,\
+        "failed to run the kubectl command"
+    last_message = ''
+
+    cmd = "kubectl get daemonset -n ingress-nginx --no-headers -o name"
+    send_a_command(ws, cmd)
+    print('kubectl command and output:\n' + last_message + '\n')
+    target = last_message.split("\r\n")[1]
+    assert "daemonset.apps/nginx-ingress-controller" in target
+    last_message = ''
+
+    ws.close()
+
+
+def test_websocket_exec_shell():
+    global last_message
+    url = 'wss://' + CATTLE_TEST_URL[8:] + \
+        '/k8s/clusters/' + namespace["cluster"].id + \
+        '/api/v1/namespaces/' + namespace["ns"] + \
+        '/pods/' + namespace["pod"].name + \
+        '/exec?container=' + namespace["pod"].containers[0].name + \
+        '&stdout=1&stdin=1&stderr=1&tty=1&command=%2Fbin%2Fsh&command=-c' \
```

Comment:
Is it possible to have the url in plaintext and then urlencode afterwards? it would make it much easier to see the arguments passed into the endpoint

Reply:
now the parameters are listed and then added to the url.